### DISCUSSION TITLE
Experiment with dagger SDK/Mage for running registry and test tasks

### DISCUSSION THREAD

**Discussion by sheldonhull** [aquaproj/aqua#2576] [Deprecated (Proposal)]
### Overview

Dagger runs containerized. Runs equally well on a windows machine, macOS, or Linux, and would allow Go based test and configuration runs without any environment impact.

### Why is the proposal needed?

I noticed the cmdx tool is used which is pretty cool (reminds me of Task).
Consider using Mage as an alternative to experiment with which would let you eliminate any core reliance on shell scripts for automation.

### Note

I'm not asking you to go do it, rather if you'd be open a to demo magefile contribution that could include automating some of the tasks that use containers and a few supporting tasks as a start.  

It's not high priority, but if interested I'd be glad to demo something.
I've used for over 2 years and found it great for automation in nearly 100% Go, while being low friction. The Dagger SDK provides another level of amazing automation capabilities that isolate the system from being impacted by the environment, perhaps improving long term maintenance.

**Comment by suzuki-shunsuke** [comment]
Thank you for your proposal.
Let me confirm if my understanding is correct.

We use cmdx, shell scripts, registry-tool, and Docker to build aqua-registry's local development workflow.

- https://github.com/suzuki-shunsuke/cmdx
- https://github.com/aquaproj/aqua-registry/blob/main/cmdx.yaml
- https://github.com/aquaproj/aqua-registry/tree/main/scripts
- https://github.com/aquaproj/registry-tool

So in my understanding, your proposal is to replace them with [Dagger](https://docs.dagger.io/) or [Mage](https://magefile.org/). Is this correct?

Indeed, current cmdx tasks and shell scripts depend on bash and environment so I wanted to migrate parts of them to registry-tool.
I'm not familiar with Dagger and Mage, but I'm interested in how they improve our experience.

BTW, we ever used [Earthly](https://earthly.dev/) but stopped using it as Earthly was a little slow and the current implementation with cmdx and Docker was faster and better experience than Earthly. https://github.com/aquaproj/aqua-registry/pull/16597

**Reply by suzuki-shunsuke** [reply]
I tried Dagger, though this is still work in progress.

- https://github.com/aquaproj/aqua-registry/pull/22777

Dagger is interesting but have an overhead so I'm not sure if we should replace existing scripts with Dagger.
